### PR TITLE
feat: add user fonts list

### PR DESCRIPTION
**Pull request by rrcwang**
#980

### CODE REVIEW

File: shared/rust/src/domain/user.rs [hunk 1]
```diff
@@ -285,3 +285,21 @@ pub struct UserColorResponse {
     /// The user's colors.
     pub colors: Vec<rgb::RGBA8>,
 }
+
+/// Request for [`CreateFont`]
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
+#[cfg_attr(feature = "backend", openapi(empty))]
```
Comment (izik1) [inline shared/rust/src/domain/user.rs:292]: drop the `openapi(empty)` if it compiles without it please. `openapi(empty)` is essentially a workaround to paperclip not understanding some types.

File: shared/rust/src/domain/user.rs [hunk 2]
```diff
@@ -285,3 +285,21 @@ pub struct UserColorResponse {
...
+pub struct UserFontNameRequest {
+    /// Name of the font to add/change.
+    pub font_name: String,
+}
+
+/// Response for [`CreateFont`]
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
+#[cfg_attr(feature = "backend", openapi(empty))]
```
Comment (izik1) [inline shared/rust/src/domain/user.rs:301]: drop the `openapi(empty)` if it compiles without it please. `openapi(empty)` is essentially a workaround to paperclip not understanding some types.

File: shared/rust/src/domain/user.rs [hunk 3]
```diff
@@ -285,3 +285,21 @@ pub struct UserColorResponse {
...
     pub colors: Vec<rgb::RGBA8>,
 }
+
+/// Request for [`CreateFont`]
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
+#[cfg_attr(feature = "backend", openapi(empty))]
+pub struct UserFontNameRequest {
+    /// Name of the font to add/change.
+    pub font_name: String,
```
Comment (izik1) [inline shared/rust/src/domain/user.rs:295]: Consider naming this `name` (same goes with the tables)

File: shared/rust/src/api/endpoints/user/fonts.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+pub struct Update;
+impl ApiEndpoint for Update {
+    type Req = UserFontNameRequest;
+    type Res = ();
+    type Err = EmptyError;
+    const PATH: &'static str = "/v1/user/me/font/{index}";
+    const METHOD: Method = Method::Patch;
+}
+
+/// Delete a user color.
```
Comment (izik1) [inline shared/rust/src/api/endpoints/user/fonts.rs:41]: Missed doc comment change?

File: backend/api/src/db/user.rs [hunk 5]
```diff
@@ -330,3 +330,134 @@ where index > $2 and user_id = $1
...
+
+    Ok(font_names.into_iter().map(|it| it.font_name).collect())
+}
+
+pub async fn delete_font(db: &sqlx::PgPool, user_id: Uuid, index: u16) -> sqlx::Result<()> {
+    let mut txn = db.begin().await?;
+
+    let _ = sqlx::query!(
+        r#"
+with delete as (
```
Comment (izik1) [inline backend/api/src/db/user.rs:433]: Can I get a double check here that the query planner doesn't optimize the locks out?

File: backend/api/migrations/20210426153720_user-font-table.sql [hunk 6]
```diff
@@ -0,0 +1,6 @@
+create table user_font (
+    user_id     uuid    not null references "user" (id) on delete cascade,
+    "index"     int2    not null,
```
Comment (izik1) [inline backend/api/migrations/20210426153720_user-font-table.sql:3]: consider adding a `check ("index" >= 0)` here, all integers in postgres are signed

**Review by izik1 — APPROVED** [review]
few nits that can be fixed in a follow up pr, generally very clean

File: backend/api/src/db/user.rs (hunk 5 above)
Comment (rrcwang) [reply backend/api/src/db/user.rs:433]: Looks to me the locks are held:
```
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 LockRows  (cost=12.35..17.71 rows=2 width=10)
   CTE delete
     ->  Delete on user_font user_font_1  (cost=0.15..8.17 rows=1 width=6)
           ->  Index Scan using user_font_user_id_index_key on user_font user_font_1  (cost=0.15..8.17 rows=1 width=6)
                 Index Cond: ((user_id = '1f241e1b-b537-493f-a230-075cb16315be'::uuid) AND (index = 1))
   ->  Bitmap Heap Scan on user_font  (cost=4.17..9.52 rows=2 width=10)
         Recheck Cond: ((user_id = '1f241e1b-b537-493f-a230-075cb16315be'::uuid) AND (index > 1))
         ->  Bitmap Index Scan on user_font_user_id_index_key  (cost=0.00..4.17 rows=2 width=0)
               Index Cond: ((user_id = '1f241e1b-b537-493f-a230-075cb16315be'::uuid) AND (index > 1))
```

**Pull request merged**